src/App.js

```diff
@@ -18,23 +18,50 @@ const withOnlineStatus = WrappedComponent =>
     constructor(props) {
       super(props);
       this.state = { isOnline: false };
+      this.quickDisconnectMS = 2000;
+      this.lastTimeOffline = null;
+      this.offlineTimer = null;
     }
-    render() {
+
+    setOffline = () => {
+      this.offlineTimer = null;
+      // Since offline notifications are deferred, we have to check duplication right before setState too, 
+      // because we cannot know isOnline state in future
+      if (!this.state.isOnline) return; // Skip duplications
+      this.setState({ isOnline: false });
+    }
+
+    onIsOnlineChange = nextIsOnline => {
       const { isOnline } = this.state;
+      const now = Date.now();
+      if (nextIsOnline && this.lastTimeOffline && now - this.lastTimeOffline < this.quickDisconnectMS) {
+        // Quick disconnect happened
+        clearTimeout(this.offlineTimer) // Cancel deferred offline notification
+        this.offlineTimer = null;
+        return; // Skip, because it's already online
+      }
+      if (nextIsOnline && !isOnline) {
+        this.setState({ isOnline: true }); // Switch to online, skipping duplications
+      } else if (!nextIsOnline && isOnline) {
+        if (this.offlineTimer) clearTimeout(this.offlineTimer); // Clear previous deferred offline switch, if exists
+        this.offlineTimer = setTimeout(this.setOffline, this.quickDisconnectMS); // Set new deferred switch to offline
+        this.lastTimeOffline = now;
+      }
+    }
```

Comment:
I decided to use a pure JS solution.  
The logic looks transparent to me

Reply:
Looks great to me. Indeed, very bulletproof and clear 👍 Just one small question:
**Input**:
`Online — Offline, Offline, Offline, Offline, Offline, ...`
**Output Notifications**:
Online. It only prints `Offline` 2 seconds after the _last_ `Offline` event of the series. 
**The output that sounds more natural to me**
Online, Offline, and then silence. I mean, I'd expect `Offline` to appear 2 seconds after the _first_ `Offline` event of the series.

Was that an understatement from us or your conscious decision?

Reply:
Sorry, I missed that during testing. 
Fixed now, and made it even more simple